r/R/atac_utils.R

```diff
@@ -379,30 +379,260 @@ call_peaks_tile <- function(fragments, chromosome_sizes, cell_groups = rep.int("
+#' as well as a shell file containing a call to MACS are written for each cell group.
+#'  - Bed files containing `chr`, `start`, and `end` coordinates of insertions are written at `<path>/input/<group>.bed.gz`.
+#'  - Shell commands to run MACS manually are written at `<path>/input/<group>.sh`.
+#'
+#' Outputs are written to an output directory with a subdirectory for each cell group.
+#' Each cell group's output directory contains a file for narrowPeaks, peaks, and summits.
+#'  - NarrowPeaks are written at `<path>/output/<group>/<group>_peaks.narrowPeak`.
+#'  - Peaks are written at `<path>/output/<group>/<group>_peaks.xls`.
+#'  - Summits are written at `<path>/output/<group>/<group>_summits.bed`.
+#'
+#' Only the narrowPeaks file is read into a tibble and returned.
+#' For more information on outputs from MACS, visit the [MACS docs](https://macs3-project.github.io/MACS/docs/callpeak.html)
+#'
+#' **Performance**:
+#' 
+#' Running on a 2600 cell dataset and taking both start and end insertions into account, written input bedfiles and MACS outputs 
+#' used 364 MB and 158 MB of space respectively.  With 4 threads, running this function end to end took 74 seconds, with 61 of those seconds spent on running MACS.
+#'
+#' **Running MACS manually**:
+#'
+#' To run MACS manually, you will first run `call_macs_peaks()` with `step="prep-inputs`. Then, manually run all of the
+#' shell scripts generated at `<path>/input/<group>.sh`. Finally, run `call_macs_peaks()` again with the same original arguments, but
+#' setting `step="read-outputs"`.
+#' @inheritParams call_peaks_tile
+#' @export
+call_macs_peaks <- function(fragments, path,
+                            cell_groups = rlang::rep_along(cellNames(fragments), "all"), effective_genome_size = 2.7e9,
+                            insertion_mode = c("start_only", "both", "end_only"),
+                            step = c("all", "prep-inputs", "run-macs", "read-outputs"),
+                            macs_executable = NULL,
+                            additional_params = "--call-summits --keep-dup all --shift -75 --extsize 150 --nomodel --nolambda",
+                            verbose = FALSE,
+                            threads = 1) {
+  assert_is(fragments, c("IterableFragments", "NULL"))
+  assert_is(cell_groups, c("character", "factor"))
+  assert_is_numeric(effective_genome_size)
+  assert_is_character(path)
+  assert_is_wholenumber(threads)
+  insertion_mode <- match.arg(insertion_mode)
+  step <- match.arg(step)
+  cell_groups <- as.factor(cell_groups)
+  levels(cell_groups) <- normalize_unique_file_names(levels(cell_groups))
+  path <- normalizePath(path)
+  path_bed_input <- paste0(path, "/input/", levels(cell_groups), ".bed.gz")
+  names(path_bed_input) <- levels(cell_groups)
+  path_macs_output <- paste0(path, "/output/", levels(cell_groups))
+  # Create paths
+  dir.create(file.path(path, "input"), showWarnings = FALSE, recursive = TRUE)
```

Comment:
```suggestion
  # Create paths
  dir.create(file.path(path, "input"), showWarnings = FALSE, recursive = TRUE)
  path <- normalizePath(path)
  path_bed_input <- paste0(path, "/input/", levels(cell_groups), ".bed.gz")
  names(path_bed_input) <- levels(cell_groups)
  path_macs_output <- paste0(path, "/output/", levels(cell_groups))
```
I think this slight re-order is better, thinking about the case where `path` is a not-yet-existing directory. `dir.create` will do the `~` expansion, but `normalizePath()` will error out on a non-existent path (or else operate with reduced efficiency if `mustWork=FALSE`)

Reply:
Agreed, I was getting a ton of warnings on `normalizePath()` on the bed section so this would make sense on `call_macs_peaks()`